src/hfdl.rs

```diff
@@ -0,0 +1,629 @@
+#[serde(deny_unknown_fields)]
+pub struct ICAOFacilityIdData {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    icao_facility_name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    icao_facility_designation: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ATCFreq {
+    pub choice: String,
+    pub data: ATCFreqData,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ATCFreqData {
+    vhf: ATCFreqDataType,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ATCFreqDataType {
+    val: f64,
+    unit: FrequencyLabel
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(try_from = "String")]
+pub enum FrequencyLabel {
+    #[default]
+    MHz,
+}
+
+impl TryFrom<String> for FrequencyLabel {
+    type Error = String;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        match value.as_str() {
+            "MHz" => Ok(FrequencyLabel::MHz),
+            _ => Err(format!("Unknown FrequencyLabel: {}", value))
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ATCUplinkMsg {
+    pub header: ATCUplinkHeader,
+    pub atc_uplink_msg_element_id: ATCUplinkMessageElementId
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ATCUplinkHeader {
+    msg_id: u16,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    msg_ref: Option<u16>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    timestamp: Option<UTCTime>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct LPDUAcarsMediaAdvisory {
+    err: bool,
+    version: u8,
+    current_link: LPDUAcarsMediaAdvisoryLink,
+    links_avail: Vec<LPDUACARSMediaAdivsoryLinksAvailble>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct LPDUAcarsMediaAdvisoryLink {
+    code: String,
+    descr: String,
+    established: bool,
+    time: UTCTime
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct LPDUACARSMediaAdivsoryLinksAvailble {
+    code: String,
+    descr: String,
+}
+
+#[derive(Serialize, Debug, Deserialize, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct SPDUorLPDUSource {
+    #[serde(rename = "type")]
+    source_type: LPDUSrcType,
+    id: u16,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    ac_info: Option<LPDUAircraftInfo>,
+}
+
+#[derive(Deserialize, Debug, Clone, Eq, PartialEq, Default)]
+#[serde(try_from = "String", deny_unknown_fields)]
+pub enum LPDUSrcType {
+    #[default]
+    Aircraft,
+    GroundStation
+}
+
+// Helper to serialize the enum back to the original string.
+impl Serialize for LPDUSrcType {
```

Comment:
I'm on the fence about this particular one...

Whilst the code is fine, could probably have a couple of functions defined for "get_source_type" or similar that do a string comparison check and return a bool/enum/whatever.

Personally I only really bother with doing this when I _want_ it to error when it hits an unknown value, because it should only ever be one of a set collection of values.

If you want it fail parsing the message if it's not `Aircraft` or `Ground station`, this is fine.

Reply:
Fixed! I turned off `cargo fmt` so it wouldn't break your code styling in this and I missed it.

Reply:
This was one of the very first things I added in as I started work on the decoder. My initial idea was to get super super specific about all of the types for future use in things like ACARS Hub. I wasn't sure the best way of doing it, and I went with this.

I'm happy to move it to be more lenient and just serialize to a string and add in methods like you've suggested, but I'm sort of partial to the idea of having it fail to deserialize if the value in the string doesn't match what we expect.

Perhaps the best compromise would be to successfully deserialize and spit out warnings like you mentioned on discord? I'm not sure. I feel like whatever method is used should ideally be applied to the entire code base, and I'm happy to follow your lead.